logsmanagement/flb_plugin.c

```diff
@@ -0,0 +1,982 @@
+            p_file_info->flb_tmp_docker_ev_metrics.ev_type[i] = 0;
+        }
+        uv_mutex_unlock(p_file_info->parser_metrics_mut);
+    } else m_assert(0, "invalid p_file_info->log_type at this point");
+
+    buff->in->timestamp = 0;
+    buff->in->text_size = 0;
+    // *buff->in->data = 0;
+
+    uv_mutex_unlock(&p_file_info->flb_tmp_buff_mut);
+}
+
+static int flb_write_to_buff_cb(void *record, size_t size, void *data){
+
+    struct File_info *p_file_info = (struct File_info *) data;
+    Circ_buff_t *buff = p_file_info->circ_buff;
+
+    msgpack_unpacked result;
+    size_t off = 0; 
+    struct flb_time tmp_time;
+    msgpack_object *x;   
+
+    /* FLB_WEB_LOG case */
+    char *syslog_severity = NULL;
+    size_t syslog_severity_size = 0;
+    char *syslog_facility = NULL;
+    size_t syslog_facility_size = 0;
+    char *syslog_timestamp = NULL;
+    size_t syslog_timestamp_size = 0;
+    char *hostname = NULL;
+    size_t hostname_size = 0;
+    char *syslog_identifier = NULL;
+    size_t syslog_identifier_size = 0;
+    char *pid = NULL;
+    size_t pid_size = 0;
+    char *message = NULL;
+    size_t message_size = 0;
+    /* FLB_WEB_LOG case end */
+
+    /* FLB_DOCKER_EV case */
+    long docker_ev_time = 0;
+    long docker_ev_timeNano = 0;
+    char *docker_ev_type = NULL;
+    size_t docker_ev_type_size = 0;
+    char *docker_ev_action = NULL;
+    size_t docker_ev_action_size = 0;
+    char *docker_ev_id = NULL;
+    size_t docker_ev_id_size = 0;
+    static struct Docker_ev_attr {
+        char **key;
+        char **val;
+        size_t *key_size; 
+        size_t *val_size;
+        int size, max_size;
+    } docker_ev_attr = {0};
+    docker_ev_attr.size = 0;
+    /* FLB_DOCKER_EV case end */
+
+    size_t new_tmp_text_size = 0;
+
+    msgpack_unpacked_init(&result);
+
+    uv_mutex_lock(&p_file_info->flb_tmp_buff_mut);
+
+    int iter = 0;
+    while (dl_msgpack_unpack_next(&result, record, size, &off) == MSGPACK_UNPACK_SUCCESS) {
+        iter++;
+        m_assert(iter == 1, "We do not expect more than one loop iteration here");
+
+        flb_time_pop_from_msgpack(&tmp_time, &result, &x);
+        if(buff->in->timestamp == 0) {
+            // printf("timestamp text_size:%zu\n", buff->in->text_size);
+            // printf("timestamp text:%s\n", buff->in->text ? buff->in->text : "NULL");
+            // fflush(stdout);
+            m_assert(buff->in->text_size == 0, "buff->in->timestamp == 0 but buff->in->text_size != 0");
+            // m_assert(buff->in->data == 0 || 
+            //         *buff->in->data == 0, "buff->in->timestamp == 0 but *buff->in->text != 0");
+
+            buff->in->timestamp = (uint64_t) tmp_time.tm.tv_sec * MS_IN_SEC + (uint64_t) tmp_time.tm.tv_nsec / (MS_IN_SEC * MS_IN_SEC);
+            m_assert(buff->in->timestamp > 1649175852000 && buff->in->timestamp < 2532788652000, "buff->in->timestamp is invalid"); // Timestamp within valid range up to 2050
```

Comment:
I understood the motive you created an upper limit, but why did you also have a lower limit?
Why is necessary to be `April 5, 2022 4:24:12 PM`? I prefer `August 30th, 1985 09:00:00 AM`. :smile: 
Why did we not set the lowest value (if it is necessary)  the day that plugin is released for the first time? Or when it was installed.
Another thing, please, convert them to `define`.

Reply:
There was a bug once where `buff->in->timestamp` was in seconds instead of milliseconds, somehow. That's why I have the lower limit. Keep in mind `m_assert` is only enabled if `LOGS_MANAG_DEBUG_LEV` is defined. 

I have changed that to make it more clear now and added comments: https://github.com/netdata/netdata/pull/13291/commits/1d32c819a58709fd14224a4e164655ab7ccec38e

Reply:
Interesting :thinking: .
We will need to dig this better before to merge. I am not convinced that we should have a small limit so young.

Reply:
Well, we can of course change the lower limit to anything else. It was arbitrarily selected.

Reply:
My concern with this kind of lower limit is the following:

Let us suppose that an user has a host with ancient logs and he wants to use netdata to understand what happened. When we set limits like this, we are creating barriers for user to test old scenarios. 

I know you can argue that users in normal situation won't keep so long, but I can contradict saying that I already worked on a company where we saved more than one year of logs. So if this company installs right now Netdata, they could not get data that are still in their hard disk with this plugin. Probably we do not want set a lower limit directly, why do we not create an option in the configuration file for users to define them? :thinking:

Reply:
The code as it stands won't collect any historical data. It will collect logs that were only generated live, when the agent is running, so this won't ever be a problem. Collecting logs generated in the past (e.g. maybe when the agent was offline) can be done in a follow up PR. 

On top of that, the user will need to have `LOGS_MANAG_DEBUG_LEV` defined to enable `m_assert`, which they shouldn't. Testing the lower limit is only for internal use :) .

Reply:
All right I got, but what will happen if the system has a wrong date time older than April 5th? :)

Reply:
Well, we'd better make sure our test / development systems do not have a wrong date time! If they _do_ have a wrong one, it's better if we catch it with an assert, otherwise my entire testing would be based on wrong assumptions (i.e. logs collected at entirely wrong timestamps).